### PR TITLE
Solrbuilder/Solrupdater improvements in prep for Solr 8

### PR DESCRIPTION
**Pull request by cdrini**
Extracted from #4337 . Various improvments to perf / error handling to solrbuilder and friends. Solr builder at this point should be considered Solr 8 only. I recommend reviewing commit by commit; easier that way.

### Testing
The most recent full solr reindex was run with this code -- excluding last commit, and the fact that it was run on an older master.

Otherwise:
- ✅ Tested solr updater works correctly in local env with Solr 3
- ✅ Tested solr updater does not cache between update_keys calls (last commit)

### Stakeholders
@mekarpeles

### CODE REVIEW

File: openlibrary/solr/update_work.py [hunk 1]
```diff
@@ -1139,6 +1141,37 @@ def toxml(self):
...
 
+
+def solr8_update(
+        reqs: List[Union[str, UpdateRequest, DeleteRequest]],
+        commit_within=60_000,
+        skip_id_check=False,
+        solr_base_url: str = None,
+) -> None:
+    """This will replace solr_update once we're fully on Solr 8.7+"""
+    req_strs = (r if type(r) == str else r.toxml() for r in reqs if r)  # type: ignore
```
Comment (mekarpeles) [inline openlibrary/solr/update_work.py:1152]:
```suggestion
    req_strs = (r if type(r) is str else r.toxml() for r in reqs if r)  # type: ignore
```

pep8

File: openlibrary/solr/update_work.py [hunk 2]
```diff
@@ -1139,6 +1141,37 @@ def toxml(self):
         if self.keys:
             return make_delete_query(self.keys)
 
+
+def solr8_update(
+        reqs: List[Union[str, UpdateRequest, DeleteRequest]],
+        commit_within=60_000,
```
Comment (mekarpeles) [inline openlibrary/solr/update_work.py:1147]:
```suggestion
        commit_within=60_000,  # ms
```

File: openlibrary/solr/update_work.py [hunk 3]
```diff
@@ -1139,6 +1141,37 @@ def toxml(self):
...
+    params = {}
+    if commit_within is not None:
+        params['commitWithin'] = commit_within
+    if skip_id_check:
+        params['overwrite'] = 'false'
+    logger.debug(f"POSTing update to {solr_base_url}/update {params}")
+    try:
+        resp = httpx.post(
+            f'{solr_base_url}/update',
+            timeout=30,  # The default timeout is silly short
```
Comment (mekarpeles) [inline openlibrary/solr/update_work.py:1166]:
```suggestion
            timeout=30,  # seconds (The default timeout is silly short)
```

File: openlibrary/solr/update_work.py [hunk 4]
```diff
@@ -1465,7 +1498,13 @@ def solr_select_work(edition_key):
...
 
 
-def update_keys(keys, commit=True, output_file=None, commit_way_later=False):
+def update_keys(keys,
+                commit=True,
+                output_file=None,
+                commit_way_later=False,
+                solr8=False,
+                skip_id_check=False,
+                update: Literal['update', 'print', 'pprint', 'quiet'] = 'update'):
```
Comment (mekarpeles) [inline openlibrary/solr/update_work.py:1507]: we're only using some of these options

File: scripts/solr_builder/solr_builder/solr_builder.py [hunk 5]
```diff
@@ -204,51 +232,93 @@ async def _get_lite_metadata(ocaids, _recur_depth=0, _max_recur_depth=3):
...
-                    'fl': 'identifier,boxid,collection',
-                    'page': 1,
-                    'output': 'json',
-                    'save': 'yes',
-                }
-            )
-
         try:
+            async with httpx.AsyncClient() as client:
+                r = await client.get(
```
Comment (mekarpeles) [inline scripts/solr_builder/solr_builder/solr_builder.py:237]: :( my thoughts are known, it's fine

**Review by mekarpeles — APPROVED** [review]
lgtm, few comments to resolve at your discretion, please merge when ready

File: openlibrary/solr/update_work.py (hunk 1 above)
Comment (cdrini) [reply openlibrary/solr/update_work.py:1152]: Going to punt on this one; I don't believe it's necessary in this case? @dhruvmanila @cclauss Do you by any chance know?

File: openlibrary/solr/update_work.py (hunk 2 above)
Comment (cdrini) [reply openlibrary/solr/update_work.py:1147]: Done

File: openlibrary/solr/update_work.py (hunk 3 above)
Comment (cdrini) [reply openlibrary/solr/update_work.py:1166]: Done

File: openlibrary/solr/update_work.py (hunk 4 above)
Comment (cdrini) [reply openlibrary/solr/update_work.py:1507]: Done

File: scripts/solr_builder/solr_builder/solr_builder.py (hunk 5 above)
Comment (cdrini) [reply scripts/solr_builder/solr_builder/solr_builder.py:237]: :P

File: openlibrary/solr/update_work.py (hunk 1 above)
Comment (cclauss) [reply openlibrary/solr/update_work.py:1156]: As described in PEP8, do not directly compare types but use `isinstance(r, str)` instead.

**Pull request merged**